src/include/alloc.h

```diff
@@ -106,19 +108,80 @@ niova_vbasic_nassigned(const struct niova_vbasic_allocator *nvba)
 }
 
 static inline int
-niova_vbasic_init(struct niova_vbasic_allocator *nvba,
-                  size_t region_size)
+niova_vbasic_init(struct niova_vbasic_allocator *nvba, size_t region_size)
 {
     if (!nvba)
         return -EINVAL;
 
     unsigned int unit_size = region_size / NIOVA_VBA_MAX_BITS;
-    if (unit_size == 0)
+
+    if (unit_size == 0 || unit_size > UINT32_MAX)
         return -EINVAL;
 
-    CONST_OVERRIDE(size_t, nvba->nvba_unit_size, unit_size);
+    CONST_OVERRIDE(uint32_t, nvba->nvba_unit_size, unit_size);
     nvba->nvba_bitmap = 0;
+    nvba->nvba_alignment = 0;
+
+    nvba->nvba_region_ptr = &nvba->nvba_region[0];
+    return 0;
+}
+
+/**
+ * To callers: Always allocate region size inclusive of alignment
+ */
+static inline int
+niova_vbasic_init_aligned(struct niova_vbasic_allocator *nvba,
+                          size_t region_size, size_t unit_size,
+                          size_t alignment)
+{
+    if (!nvba)
+        return -EINVAL;
+
+    if (region_size == 0)
+        return -ENODATA;
+
+    /* alignment is power of 2 */
+    if (alignment == 0 || (alignment & (alignment - 1)))
+        return -EDOM;
+
+    unit_size = (unit_size == 0) ? region_size / NIOVA_VBA_MAX_BITS : unit_size;
 
+    if (unit_size == 0 || unit_size > UINT32_MAX)
+        return -EINVAL;
+
+    if (alignment > unit_size)
+        return -EOVERFLOW;
+
+    /* Unit size is multiple of alignment */
+    if (unit_size & (alignment - 1))
+        return -EDOM;
+
+    if (region_size < unit_size)
```

Comment:
On line 180 and 181 we may reduce the effective region size.  Should we not make that reduction calculation prior this check (and the one line 147)?  IOW, I think the adjusted size should be calculated at or near the top of this function